# SCRAPING PÁGINA WEB: AUTOHERO
(Coches de segunda mano)

Hemos realizado un Scraping de la pagina, desde el entorno de Python "Pychar" y hemos obtenido 4 archivos csv que corresponden a coches de segunda mano previamente filtrando los menores a 120000 kms y con menos de 12 años de antiguedad:

In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime

#1 . Archivos csv obtenidos mediante scraping:




In [ ]:
df1 = pd.read_csv("CSV_Scraping_autohero.csv")
df2 = pd.read_csv("CSV_Scraping_autohero2.csv")
df3 = pd.read_csv("CSV_Scraping_autohero3.csv")
df4 = pd.read_csv("CSV_Scraping_autohero4.csv")

In [ ]:
# vemos los archivos
print(f"__________________________________________________________________________________\n\n \
      df1: -----> \n\n {df1.head(3)}\n\n")
print(f"__________________________________________________________________________________\n\n \
      df2: -----> \n\n {df2.head(3)}\n\n")
print(f"__________________________________________________________________________________\n\n \
      df3: -----> \n\n {df3.head(3)}\n\n")
print(f"__________________________________________________________________________________\n\n \
      df4: -----> \n\n {df4.head(3)}\n\n__________________________________________________________________________________")

__________________________________________________________________________________

       df1: -----> 

            Marca         Modelo        Version    Precio Combustible      Año  \
0   Seat Ateca,    Seat Ateca,   1.5 TSI Style  28.399 €    Gasolina  12.2021   
1  Hyundai i10,   Hyundai i10,     1.0 Essence       NaN    Gasolina  09.2022   
2  Kia Picanto,   Kia Picanto,     1.0 Concept       NaN    Gasolina  06.2020   

       Kms         Potencia Puertas                  Cambio   Color  
0   330 km  150 CV / 110 kW       5  Cambio tipo automático    Gris  
1   284 km    67 CV / 49 kW       5      Cambio tipo manual    Azul  
2  1597 km    67 CV / 49 kW       5      Cambio tipo manual  Blanco  


__________________________________________________________________________________

       df2: -----> 

                        Marca                     Modelo        Version  \
0         Dacia Sandero II,          Dacia Sandero II,    1.2 Ambiance   
1          Chevrolet Cruze,      

#2 . Unimos los archivos csv:
Unimos los 4 archivos csv en uno para trabajar comodamente:

Cada uno de los archivos contiene una cabecera y solo queremos la del primer archivo.

In [ ]:
# Función para unir los archivos csv en uno solo.

def concatenar_lista_archivos(lista_archivos):
  global df
  df = pd.read_csv(lista_archivos[0])

  for i in range (1, len(lista_archivos)):
    if i <= len(lista_archivos):
      df1 = pd.read_csv(lista_archivos[i])
      df = pd.concat([df,df1]) # concatenamos un nuevo archivo al conjunto ya concatenado y reiteramos hasta completarlo.
  return df


archivos = ["CSV_Scraping_autohero.csv", "CSV_Scraping_autohero2.csv", "CSV_Scraping_autohero3.csv", "CSV_Scraping_autohero4.csv"]
concatenar_lista_archivos(archivos)

Marca                  Modelo  \
0              Seat Ateca,             Seat Ateca,    
1             Hyundai i10,            Hyundai i10,    
2             Kia Picanto,            Kia Picanto,    
3              Kia Stonic,             Kia Stonic,    
4     Citroen C3 Aircross,    Citroen C3 Aircross,    
..                      ...                     ...   
496              Fiat 500,               Fiat 500,    
497           Peugeot 208,            Peugeot 208,    
498  Renault Grand Scenic,   Renault Grand Scenic,    
499               Kia Rio,                Kia Rio,    
500   Nissan NV200 Evalia,    Nissan NV200 Evalia,    

                            Version    Precio Combustible      Año        Kms  \
0                     1.5 TSI Style  28.399 €    Gasolina  12.2021     330 km   
1                       1.0 Essence       NaN    Gasolina  09.2022     284 km   
2                       1.0 Concept       NaN    Gasolina  06.2020    1597 km   
3    1.0 TGDI Mild-Hybrid iMT Drive       NaN    Gasolina  04.2021    1673 km   
4             1.2 PureTech C-Series  17.099 €    Gasolina  09.2020    1695 km   
..                              ...       ...         ...      ...        ...   
496                           1.2 S  12.999 €    Gasolina   5.2018  61.360 km   
497              1.2 PureTech Style   9.799 €    Gasolina   6.2017  61.384 km   
498                     1.3 TCe Zen  18.499 €    Gasolina  10.2018  61.395 km   
499    1.2 CVVT Eco-Dynamics X-Tech   8.599 €    Gasolina  12.2015  61.549 km   
500                 1.5 dCi Comfort  15.099 €      Diésel  10.2018  61.706 km   

            Potencia Puertas                  Cambio   Color  
0    150 CV / 110 kW       5  Cambio tipo automático    Gris  
1      67 CV / 49 kW       5      Cambio tipo manual    Azul  
2      67 CV / 49 kW       5      Cambio tipo manual  Blanco  
3     101 CV / 74 kW       5      Cambio tipo manual    Azul  
4     110 CV / 81 kW       5      Cambio tipo manual    Gris  
..               ...     ...                     ...     ...  
496    70 CV / 51 kW       3  Cambio tipo automático   Negro  
497    82 CV / 60 kW       5      Cambio tipo manual    Gris  
498  140 CV / 103 kW       5      Cambio tipo manual  Marrón  
499    85 CV / 62 kW       5      Cambio tipo manual  Blanco  
500   110 CV / 81 kW       4      Cambio tipo manual  Blanco  

[2506 rows x 11 columns]

## Cabecera en el centro del archivo 1, lo eliminamos.
En el centro del primer archivo se encuentra una cabecera, resultado de unir archivos durante el scraping, lo eliminamos.

In [ ]:
# Buscamos donde se encuentra esta fila con los nombres de las columnas:
df.loc[df["Año"] == "Año"]

Marca  Modelo  Version  Precio  Combustible  Año  Kms  Potencia  Puertas  \
641  Marca  Modelo  Version  Precio  Combustible  Año  Kms  Potencia  Puertas   

     Cambio  Color  
641  Cambio  Color

In [ ]:
# borramos:
df.drop([641], axis=0, inplace=True)
# y comprobamos:
df.loc[df["Año"] == "Año"]

Empty DataFrame
Columns: [Marca, Modelo, Version, Precio, Combustible, Año, Kms, Potencia, Puertas, Cambio, Color]
Index: []

#3 .Limpieza del df.

## Nulos

### Total de nulos:

In [ ]:
# donde se encuentran los nulos:
print("Nulos por columna:")
print()
df.isnull().sum()

Nulos por columna:



Marca            0
Modelo           0
Version          0
Precio         334
Combustible      0
Año              0
Kms              0
Potencia         0
Puertas          5
Cambio           0
Color            1
dtype: int64

In [ ]:
#Total de nulos:
print(f"Total nulos: {df.isnull().sum().sum()} valores")

Total nulos: 340 valores


### Eliminamos nulos:
Los nulos se encuentran principalmente en el precio. Los eliminamos puesto que no tiene sentido utilizar estos coches si no tienen precio.

In [ ]:
# eliminamos los nulos:

df = df.dropna()

print("Nulos por columna:")
print()
df.isnull().sum()

Nulos por columna:



Marca          0
Modelo         0
Version        0
Precio         0
Combustible    0
Año            0
Kms            0
Potencia       0
Puertas        0
Cambio         0
Color          0
dtype: int64

## Columna "Distrubuidor":
Añadimos una columna nueva para saber de qué página web vienen las diferentes ofertas. O de que scraping proceden las ofertas.

In [ ]:
df["Distribuidor"] = "Autohero.com"

df.head(3)

<ipython-input-79-1dcc89657793>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Distribuidor"] = "Autohero.com"


Marca                 Modelo                Version  \
0           Seat Ateca,            Seat Ateca,           1.5 TSI Style   
4  Citroen C3 Aircross,   Citroen C3 Aircross,   1.2 PureTech C-Series   
5           Seat Arona,            Seat Arona,       1.0 TSI Xperience   

     Precio Combustible      Año      Kms         Potencia Puertas  \
0  28.399 €    Gasolina  12.2021   330 km  150 CV / 110 kW       5   
4  17.099 €    Gasolina  09.2020  1695 km   110 CV / 81 kW       5   
5  21.599 €    Gasolina  01.2023  1771 km   110 CV / 81 kW       5   

                   Cambio Color  Distribuidor  
0  Cambio tipo automático  Gris  Autohero.com  
4      Cambio tipo manual  Gris  Autohero.com  
5      Cambio tipo manual  Rojo  Autohero.com

## Limpiar valores de columnas:
Vamos a limpiar los valores de las diferentes columnas, antes de proceder a cambiar los tipos de datos que contienen.

Ejemplo:

Columna "Potencia" (150 CV / 110 kW) ---> (150)

Primero vamos a ver que valores existen en las columnas, los únicos.

In [ ]:
# Valores únicos por columna:

for x in list(df.columns):
  print(f" Únicos  de columna \"{x.upper()}\": \n\n {df[x].unique()}")
  print("\n""\n")

 Únicos  de columna "MARCA": 

 ['Seat Ateca, ' 'Citroen C3 Aircross, ' 'Seat Arona, ' 'Fiat Panda, '
 'Hyundai Bayon, ' 'Opel Corsa, ' 'Peugeot 208, ' 'Hyundai i20, '
 'Hyundai Kona, ' 'Volkswagen T-Roc, ' 'Kia Picanto, ' 'MINI MINI, '
 'Seat León, ' 'Dacia Sandero II, ' 'Kia XCeed, ' 'Citroen C4, '
 'Renault Clio, ' 'Mazda 2, ' 'Fiat Punto, ' 'Fiat Tipo, '
 'Nissan Micra, ' 'Renault Arkana, ' 'Kia Stonic, ' 'Renault Twingo, '
 'Toyota Yaris, ' 'Abarth 595, ' 'Fiat 500C, ' 'Ford Transit Connect, '
 'Jeep Renegade, ' 'Seat Ibiza, ' 'Hyundai Tucson, ' "Kia cee'd, "
 'Volvo XC 40, ' 'Ford Ecosport, ' 'Dacia Sandero III, ' 'Skoda Kodiaq, '
 'Toyota Corolla, ' 'Subaru Trezia, ' 'Volkswagen Polo, ' 'Citroen C 3, '
 'BMW Serie 1, ' 'Citroen C-Elysée, ' 'Mercedes-Benz Clase CLA, '
 'Dacia Duster, ' 'Fiat 500X, ' 'Citroen C4 Cactus, ' 'Kia Sportage, '
 'Hyundai i30, ' 'Ford Puma, ' 'Ford Focus, ' 'Renault Megane, '
 'Peugeot 108, ' 'Ford Tourneo Courier, ' 'Nissan Qashqai, '
 'Peugeot 2008, ' 

In [ ]:
# Para ver que cambios debemos hacer a los valores de las columnas, mostramos el primer valor de las columnas:

for x in df.columns:
  print(f" Columna \"{x.upper()}\":\n{df[x][0]}") # imprime 3 valores porque le estoy indicando indice con el "0" y hay 3 índices con valor cero.
  print()


 Columna "MARCA":
0          Seat Ateca, 
0    Dacia Sandero II, 
0      Peugeot Rifter, 
Name: Marca, dtype: object

 Columna "MODELO":
0          Seat Ateca, 
0    Dacia Sandero II, 
0      Peugeot Rifter, 
Name: Modelo, dtype: object

 Columna "VERSION":
0          1.5 TSI Style
0           1.2 Ambiance
0    1.2 e-THP Active L2
Name: Version, dtype: object

 Columna "PRECIO":
0    28.399 €
0     7.199 €
0    22.499 €
Name: Precio, dtype: object

 Columna "COMBUSTIBLE":
0    Gasolina
0    Gasolina
0    Gasolina
Name: Combustible, dtype: object

 Columna "AÑO":
0    12.2021
0     3.2015
0     3.2019
Name: Año, dtype: object

 Columna "KMS":
0       330 km
0    70.363 km
0    72.288 km
Name: Kms, dtype: object

 Columna "POTENCIA":
0    150 CV / 110 kW
0      75 CV / 55 kW
0     110 CV / 81 kW
Name: Potencia, dtype: object

 Columna "PUERTAS":
0      5
0    5.0
0      5
Name: Puertas, dtype: object

 Columna "CAMBIO":
0    Cambio tipo automático
0        Cambio tipo manual
0        Cam

In [ ]:
df.head(5)

Marca                 Modelo                    Version  \
0           Seat Ateca,            Seat Ateca,               1.5 TSI Style   
4  Citroen C3 Aircross,   Citroen C3 Aircross,       1.2 PureTech C-Series   
5           Seat Arona,            Seat Arona,           1.0 TSI Xperience   
6           Fiat Panda,            Fiat Panda,   1.0 Mild-Hybrid City Life   
7        Hyundai Bayon,         Hyundai Bayon,             1.2 Essence 2WD   

     Precio Combustible      Año      Kms         Potencia Puertas  \
0  28.399 €    Gasolina  12.2021   330 km  150 CV / 110 kW       5   
4  17.099 €    Gasolina  09.2020  1695 km   110 CV / 81 kW       5   
5  21.599 €    Gasolina  01.2023  1771 km   110 CV / 81 kW       5   
6  13.599 €    Gasolina  09.2022  2019 km    70 CV / 51 kW       5   
7  17.199 €    Gasolina  07.2022  2794 km    84 CV / 62 kW       5   

                   Cambio   Color  Distribuidor  
0  Cambio tipo automático    Gris  Autohero.com  
4      Cambio tipo manual    Gris  Autohero.com  
5      Cambio tipo manual    Rojo  Autohero.com  
6      Cambio tipo manual    Gris  Autohero.com  
7      Cambio tipo manual  Blanco  Autohero.com

Cambiamos diferentes valores de las columnas a sus valores correctos:

In [ ]:
# Cambiamos los diferentes valores de las columnas a sus valores correctos, limpios para analizar:


df["Marca"] = df["Marca"].map(lambda x: x.replace(",","").strip().split(" ")[0])

df["Modelo"] = df["Modelo"].map(lambda x: x.strip().replace(",", "").split(" ")[1:])
df["Modelo"] = df["Modelo"].map(lambda x: " ".join(x))

df["Version"].map(lambda x: x.strip())

# 'Precio' tiene como ejemplo de valores esto: '28.399\xa0€'
# Limpiamos todo esto para quedarnos solo con "28399"
df["Precio"] = df["Precio"].map(lambda x: x.replace("\xa0€","").replace(".","").strip())

df["Combustible"] = df["Combustible"].map(lambda x: x.strip())

#Columna "Año" es algo más enrevesado, lo hacemos más abajo, aparte.

#Columna 'Kms' tiene como ejemplo de valor: '61.395 km', vamos a quedarnos con: '61395'
df["Kms"] = df["Kms"].apply(lambda x: x.strip().split(" ")[0].replace('.',''))

df["Potencia"] = df["Potencia"].apply(lambda x: x.strip().split(" ")[0])

# Columna "Puertas" vamos a unir los valores "2" y "4" con los de "3" y "5", respectivamente.
# Los valores únicos de "Puertas" son ['5', '3', '2', '4', '5.0', '4.0', '2.0', '3.0'].
# Antes de cambiar valores, lo pasamos a cadena todo (aunque ya está como objeto) porque me da error debido a los .0 por alguna razón.
df["Puertas"] = df["Puertas"].astype(str)
df['Puertas'] = df['Puertas'].apply(lambda x: x.replace('.0','')) # quitamos los '.0' de algunos valores.

df["Puertas"].loc[df[df["Puertas"] == '2'].index] = '3'
df["Puertas"].loc[df[df["Puertas"] == '4'].index] = '5'

#Columna "Cambio" tiene los valores únicos:'Cambio tipo automático','Cambio tipo manual','Semi-automático','Doble embrague'.
#Vamos a convertirlos a: "automático, manual, semi-automático y doble embrague"
def convertir_cambio(cambio):
  if cambio == "Cambio tipo automático":
    cambio = "Automático"
  elif cambio == "Cambio tipo manual":
    cambio = "Manual"
  elif cambio == "Semi-automático":
    cambio = "Semi-automático"
  elif cambio == "Doble embrague":
    cambio = "Doble embrague"
  else: print("Algo salió mal!, debe de haber algún espacio en blanco entre las cadenas")
  return cambio
df["Cambio"] = df["Cambio"].apply(lambda x: x.strip())
df["Cambio"] = df["Cambio"].apply(lambda x: convertir_cambio(x))
df["Cambio"]

#Columna "Color" está correcta aunque vamos a limpiarle por si tuviera algún espacio en blanco escondido
df["Color"] = df["Color"].apply(lambda x: x.strip())

#Columna "Distribuidor" está correcta puesto que la creamos nosotros mismos.

<ipython-input-83-d8388932a9e0>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Marca"] = df["Marca"].map(lambda x: x.replace(",","").strip().split(" ")[0])
<ipython-input-83-d8388932a9e0>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Modelo"] = df["Modelo"].map(lambda x: x.strip().replace(",", "").split(" ")[1:])
<ipython-input-83-d8388932a9e0>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead



### Columna "Año":

In [ ]:
df["Año"].head(2)

0    12.2021
4    09.2020
Name: Año, dtype: object

In [ ]:
# Convertimos en string y nos quedamos con el año

df["Año"] = df["Año"].astype(str)

df["Año"] = df["Año"].apply(lambda x: x.split(".")[-1])
df["Año"].head()

<ipython-input-85-85230bbef415>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Año"] = df["Año"].astype(str)
<ipython-input-85-85230bbef415>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Año"] = df["Año"].apply(lambda x: x.split(".")[-1])


0    2021
4    2020
5    2023
6    2022
7    2022
Name: Año, dtype: object

In [ ]:
# Le añadimos un "0" a los años que al pasarlos a cadena, se les quitó el cero de la derecha.
def añadir_cero(string):
  if len(string) < 4:
    string = string + "0"
  return string

df["Año"] = df["Año"].apply(lambda x: añadir_cero(x))
df["Año"].head()

<ipython-input-86-65b95eb76ec5>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Año"] = df["Año"].apply(lambda x: añadir_cero(x))


0    2021
4    2020
5    2023
6    2022
7    2022
Name: Año, dtype: object

In [ ]:
#Comprobamos que los valores estén todos correctos y que no se haya modificado ningún otro al cambio a string:

df["Año"].unique()

array(['2021', '2020', '2023', '2022', '2019', '2018', '2017', '2012',
       '2013', '2016', '2014', '2015', '2011'], dtype=object)

Comprobamos que todas las columnas estén correctas:

In [ ]:
# Valores únicos por columna:

for x in list(df.columns):
  print(f" Únicos  de columna \"{x.upper()}\": \n\n {df[x].unique()}")
  print("\n""\n")

 Únicos  de columna "MARCA": 

 ['Seat' 'Citroen' 'Fiat' 'Hyundai' 'Opel' 'Peugeot' 'Volkswagen' 'Kia'
 'MINI' 'Dacia' 'Renault' 'Mazda' 'Nissan' 'Toyota' 'Abarth' 'Ford' 'Jeep'
 'Volvo' 'Skoda' 'Subaru' 'BMW' 'Mercedes-Benz' 'Cupra' 'Chevrolet'
 'Suzuki' 'Mitsubishi' 'Land' 'Honda' 'Smart' 'Audi' 'Ssangyong' 'Lexus'
 'DS' 'Lancia' 'Alfa' 'Jaguar' 'MG']



 Únicos  de columna "MODELO": 

 ['Ateca' 'C3 Aircross' 'Arona' 'Panda' 'Bayon' 'Corsa' '208' 'i20' 'Kona'
 'T-Roc' 'Picanto' 'MINI' 'León' 'Sandero II' 'XCeed' 'C4' 'Clio' '2'
 'Punto' 'Tipo' 'Micra' 'Arkana' 'Stonic' 'Twingo' 'Yaris' '595' '500C'
 'Transit Connect' 'Renegade' 'Ibiza' 'Tucson' "cee'd" 'XC 40' 'Ecosport'
 'Sandero III' 'Kodiaq' 'Corolla' 'Trezia' 'Polo' 'C 3' 'Serie 1'
 'C-Elysée' 'Clase CLA' 'Duster' '500X' 'C4 Cactus' 'Sportage' 'i30'
 'Puma' 'Focus' 'Megane' '108' 'Tourneo Courier' 'Qashqai' '2008'
 'Clase B' '500' 'CX 3' 'Superb' 'Duster II' 'C-HR' 'Formentor' 'Aveo'
 'SX4 S-Cross' 'Captur' 'Compass' 'Tarraco' 'A

## Transformación columnas, tipo de datos:

Transformación de datos a valores correctos.

In [ ]:
df.dtypes

Marca           object
Modelo          object
Version         object
Precio          object
Combustible     object
Año             object
Kms             object
Potencia        object
Puertas         object
Cambio          object
Color           object
Distribuidor    object
dtype: object

Todas son objetos, vamos a transformar las que no deben serlo a sus tipos correspondientes:

1. "Precio"(object) -------> int64.
2. "Año"(object) -----------> datetime[ns]
3. "Kms"(object) ----------> Float64.
4. "Potencia"(object)----> Int64

In [ ]:
df.dtypes

Marca           object
Modelo          object
Version         object
Precio          object
Combustible     object
Año             object
Kms             object
Potencia        object
Puertas         object
Cambio          object
Color           object
Distribuidor    object
dtype: object

In [ ]:
# Convertimos a la vez todas las columnas a su tipo correspondiente:
# Para la columna "Año" lo hacemos diferente, más abajo.
# Para convertir el precio a un int(), nos encontramos con que hay precios que contienen un "." como separador de miles.
# Limpiamos previamente este "." y despues convertimos:

df['Precio'] = df['Precio'].apply(lambda x: x.replace(".", ""))

df = df.astype({"Precio":"int64", "Potencia":"Int64", "Kms":"Float64"})
df["Año"] = pd.to_datetime(df["Año"], format='%Y')


<ipython-input-91-782453c22a54>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Precio'] = df['Precio'].apply(lambda x: x.replace(".", ""))


In [ ]:
df['Año'].unique()

array(['2021-01-01T00:00:00.000000000', '2020-01-01T00:00:00.000000000',
       '2023-01-01T00:00:00.000000000', '2022-01-01T00:00:00.000000000',
       '2019-01-01T00:00:00.000000000', '2018-01-01T00:00:00.000000000',
       '2017-01-01T00:00:00.000000000', '2012-01-01T00:00:00.000000000',
       '2013-01-01T00:00:00.000000000', '2016-01-01T00:00:00.000000000',
       '2014-01-01T00:00:00.000000000', '2015-01-01T00:00:00.000000000',
       '2011-01-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [ ]:
df.dtypes

Marca                   object
Modelo                  object
Version                 object
Precio                   int64
Combustible             object
Año             datetime64[ns]
Kms                    Float64
Potencia                 Int64
Puertas                 object
Cambio                  object
Color                   object
Distribuidor            object
dtype: object

## Eliminamos duplicados:

In [ ]:
len(df)

2166

In [ ]:
# Vemos si exiten filas duplicadas y cuántas:

print(f"Existen {df.duplicated(keep='first').sum()} filas duplicadas, teniendo en cuenta todas las columnas.")
print()
df[df.duplicated()].head(5) # Mostramos los duplicados. No se toma en cuenta la primera aparición de estos.

Existen 749 filas duplicadas, teniendo en cuenta todas las columnas.



Marca      Modelo                   Version  Precio Combustible  \
87       Dacia  Sandero II           0.9 TCe Stepway   10299    Gasolina   
88      Subaru      Trezia       1.4 Diesel Limited     9199      Diésel   
90  Volkswagen        Polo           1.0 TSI Advance   15299    Gasolina   
91     Citroen         C 3            PureTech Shine   15999    Gasolina   
92     Peugeot         208  1.2 PureTech Active Pack   15199    Gasolina   

          Año      Kms  Potencia Puertas  Cambio     Color  Distribuidor  
87 2017-01-01  14281.0        90       5  Manual    Blanco  Autohero.com  
88 2012-01-01  14432.0        90       5  Manual     Plata  Autohero.com  
90 2019-01-01  14521.0        95       5  Manual    Blanco  Autohero.com  
91 2021-01-01  14858.0       110       5  Manual     Negro  Autohero.com  
92 2021-01-01  14859.0        75       5  Manual  Amarillo  Autohero.com

In [ ]:
# Eliminamos las filas duplicadas:

df.drop_duplicates(subset=None, keep='first', inplace=True, ignore_index=False) # argumentos por defecto tambien funciona.

In [ ]:
# Comprobamos:
len(df)

1417

Finalmente tenemos el siguiente número de ofertas de coches (filas):

In [ ]:
print(f"Contamos con {'{:,d}'.format(len(df))} ofertas de coches de segunda mano.")

Contamos con 1,417 ofertas de coches de segunda mano.


# 4 Guardar df como .csv sin índices.

In [ ]:
df.to_csv('CSV_Scraping_autohero_LIMPIO.csv', index=False)

In [ ]:
df2 = pd.read_csv('CSV_Scraping_autohero_LIMPIO.csv')
df2

Marca        Modelo                       Version  Precio Combustible  \
0        Seat         Ateca                 1.5 TSI Style   28399    Gasolina   
1     Citroen   C3 Aircross         1.2 PureTech C-Series   17099    Gasolina   
2        Seat         Arona             1.0 TSI Xperience   21599    Gasolina   
3        Fiat         Panda     1.0 Mild-Hybrid City Life   13599    Gasolina   
4     Hyundai         Bayon               1.2 Essence 2WD   17199    Gasolina   
...       ...           ...                           ...     ...         ...   
2161     Fiat           500                         1.2 S   12999    Gasolina   
2162  Peugeot           208            1.2 PureTech Style    9799    Gasolina   
2163  Renault  Grand Scenic                   1.3 TCe Zen   18499    Gasolina   
2164      Kia           Rio  1.2 CVVT Eco-Dynamics X-Tech    8599    Gasolina   
2165   Nissan  NV200 Evalia               1.5 dCi Comfort   15099      Diésel   

             Año      Kms  Potencia  Puertas      Cambio   Color  Distribuidor  
0     2021-01-01    330.0       150        5  Automático    Gris  Autohero.com  
1     2020-01-01   1695.0       110        5      Manual    Gris  Autohero.com  
2     2023-01-01   1771.0       110        3      Manual    Rojo  Autohero.com  
3     2022-01-01   2019.0        70        5      Manual    Gris  Autohero.com  
4     2022-01-01   2794.0        84        5      Manual  Blanco  Autohero.com  
...          ...      ...       ...      ...         ...     ...           ...  
2161  2018-01-01  61360.0        70        3  Automático   Negro  Autohero.com  
2162  2017-01-01  61384.0        82        5      Manual    Gris  Autohero.com  
2163  2018-01-01  61395.0       140        5      Manual  Marrón  Autohero.com  
2164  2015-01-01  61549.0        85        5      Manual  Blanco  Autohero.com  
2165  2018-01-01  61706.0       110        5      Manual  Blanco  Autohero.com  

[2166 rows x 12 columns]